### PR TITLE
HPCC-14861 Convert inline embeds to functional form

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday gavin.halliday@lexisnexis.com

### CODE REVIEW

**Comment by ghalliday** [comment]
@shamser please review (since you are currently investigating this area)

File: ecl/hqlcpp/hqlttcpp.cpp [hunk 1]
```diff
@@ -2089,6 +2089,25 @@ IHqlExpression * ThorHqlTransformer::createTransformed(IHqlExpression * expr)
...
+            HqlExprArray args;
+            args.append(*LINK(expr));
+            if (expr->hasAttribute(languageAtom))
+            {
+                args.append(*createAttribute(contextAtom));
+                if (expr->isDatarow())
+                    args.append(*createAttribute(allocatorAtom));
+            }
+            OwnedHqlExpr body = createWrapper(no_outofline, expr->queryType(), args);
+            IHqlExpression * formals = createValue(no_sortlist, makeSortListType(NULL));
```
Comment (shamser) [inline ecl/hqlcpp/hqlttcpp.cpp]: @ghalliday Why is is the argument 'sortlist' type?

File: ecl/hqlcpp/hqlttcpp.cpp [hunk 2]
```diff
@@ -2089,6 +2089,25 @@ IHqlExpression * ThorHqlTransformer::createTransformed(IHqlExpression * expr)
...
+    case no_embedbody:
+        {
+            //Convert all definitions of non functional embeds to implicit functions
+            StringBuffer funcname;
+            funcname.append("userx").append(++implicitFunctionId);
+            HqlExprArray args;
+            args.append(*LINK(expr));
+            if (expr->hasAttribute(languageAtom))
+            {
+                args.append(*createAttribute(contextAtom));
```
Comment (shamser) [inline ecl/hqlcpp/hqlttcpp.cpp]: @ghalliday is the context atom necessary?  when would it be used?

Comment (ghalliday) [inline ecl/hqlcpp/hqlttcpp.cpp]: yes.  This code was adapted from code in hqlgram2.cpp.  The contextAtom ensure that the ICodeContext is passed to the generated function.  The generated code for non-c++ functions uses the code context in the code that is generated.

File: ecl/hqlcpp/hqlttcpp.cpp (hunk 1 above)
Comment (ghalliday) [inline ecl/hqlcpp/hqlttcpp.cpp]: A sortListType represents a heterogeneous list.  The formal parameters for a function are represented in this way - see hqlgram2.cpp(~9310)

**Comment by ghalliday** [comment]
@shamser

**Comment by shamser** [comment]
@ghalliday @richardkchapman Changes good.  Ready to merge.

**Pull request merged**